state-sync/storage-service/server/src/subscription.rs

```diff
@@ -0,0 +1,960 @@
+        }
+
+        // Verify that the subscription request index is valid
+        let subscription_request_index = subscription_request.subscription_stream_index();
+        if subscription_request_index < self.next_index_to_serve {
+            return Err((
+                Error::InvalidRequest(format!(
+                    "The subscription request index is too low! Next index to serve: {:?}, found: {:?}",
+                    self.next_index_to_serve, subscription_request_index
+                )),
+                subscription_request,
+            ));
+        }
+
+        // Verify that the request known version and epoch (at start time) are valid
+        let request_known_version = subscription_request.highest_known_version_at_stream_start();
+        let request_known_epoch = subscription_request.highest_known_epoch_at_stream_start();
+        if let Some(first_pending_request) = self.first_pending_request() {
+            let pending_known_version =
+                first_pending_request.highest_known_version_at_stream_start();
+            let pending_known_epoch = first_pending_request.highest_known_epoch_at_stream_start();
+
+            // Verify the new and pending requests have the same known version and epoch
+            if (request_known_version != pending_known_version)
+                || (request_known_epoch != pending_known_epoch)
+            {
+                return Err((
+                    Error::InvalidRequest(format!(
+                        "The subscription request known version and epoch are invalid! Expected: ({:?}), found: ({:?})",
+                        (pending_known_version, pending_known_epoch), (request_known_version, request_known_epoch)
+                    )),
+                    subscription_request,
+                ));
+            }
+        }
+
+        // Verify that the number of active subscriptions respects the maximum
+        let max_num_active_subscriptions =
+            storage_service_config.max_num_active_subscriptions as usize;
+        if self.pending_subscription_requests.len() >= max_num_active_subscriptions {
+            return Err((
+                Error::InvalidRequest(format!(
+                    "The maximum number of active subscriptions has been reached! Max: {:?}, found: {:?}",
+                    max_num_active_subscriptions, self.pending_subscription_requests.len()
+                )),
+                subscription_request,
+            ));
+        }
+
+        // Insert the subscription request into the pending requests
+        let existing_request = self.pending_subscription_requests.insert(
+            subscription_request.subscription_stream_index(),
+            subscription_request,
+        );
+
+        // Refresh the last stream update time
+        self.refresh_last_stream_update_time();
+
+        // If a pending request already existed, return the previous request to the caller
+        if let Some(existing_request) = existing_request {
```

Comment:
I wonder if clients might want to renew their subscription, in which case instead of an error maybe just the request time is updated?

But no action required I think --  just wanted to raise this. If this does become necessary, it can go in with client changes.

Reply:
Hmm... I'm on the fence 😄 Technically, the stream is renewed each time a request is added, or a response is sent. I'm not sure it makes sense for the client to send a message to renew it, but not want any data back. 🤔 Are you thinking of a specific usecase?